From 1a049dfd66128925352ea1db2abd1785b550d800 Mon Sep 17 00:00:00 2001
From: Richard Liu <xuegang.liu@nxp.com>
Date: Wed, 25 Jul 2018 14:02:26 +0800
Subject: [PATCH 4444/5242] MA-12227 [#imx-848] Refine dma-buf metadata
 setting

commit  bc37ec5514b8451e91ea8f86d1de7457daa50261 from
https://source.codeaurora.org/external/imx/linux-imx.git

Refine dma-buf metadata setting to support framebuffer compression

Change-Id: Ia0ab9224d789c4b620e33a6f1c28483025a136ce
Signed-off-by: Richard Liu <xuegang.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c |   13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
index ba3347a..2d52977 100644
--- a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
+++ b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel.c
@@ -1691,15 +1691,23 @@
     else
     {
         nodeObj->metadata.ts_fd             = Interface->u.SetVidMemMetadata.ts_fd;
-        if (nodeObj->metadata.ts_fd > 0)
+
+        if (nodeObj->metadata.ts_fd >= 0)
         {
             nodeObj->metadata.ts_dma_buf    = dma_buf_get(nodeObj->metadata.ts_fd);
+
             if (IS_ERR(nodeObj->metadata.ts_dma_buf))
             {
                 gcmkONERROR(gcvSTATUS_NOT_FOUND);
             }
+
             dma_buf_put(nodeObj->metadata.ts_dma_buf);
         }
+        else
+        {
+            nodeObj->metadata.ts_dma_buf    = NULL;
+        }
+
         nodeObj->metadata.fc_enabled        = Interface->u.SetVidMemMetadata.fc_enabled;
         nodeObj->metadata.fc_value          = Interface->u.SetVidMemMetadata.fc_value;
         nodeObj->metadata.fc_value_upper    = Interface->u.SetVidMemMetadata.fc_value_upper;
@@ -1707,9 +1715,8 @@
         nodeObj->metadata.compress_format   = Interface->u.SetVidMemMetadata.compress_format;
     }
 
-    gcmkFOOTER();
-
 OnError:
+    gcmkFOOTER();
     return status;
 }
 
-- 
1.7.9.5

